### DISCUSSION TITLE
Dynamic GitTree

### DISCUSSION THREAD

**Discussion by peterjwilson** [Kitware/trame#484] [Trame/VTK] [answered]
Hi All,

For each imported STL mesh (working fine), I want to add a new child GitTree entry under my first GitTree heading 'Meshes'. The Trame tutorials mainly presented 'static' UI layouts, whereas I am hoping to have the GitTree update dynamically. Any suggestions on how to deal with this?

**Comment by jourdain** [comment]
I'm not sure I understand what you mean. If you dynamically update the state of the sources of the git tree, it will be dynamically updated. This is its primary usage in any of the apps using it. You can find an example of it when using ParaView as a backend [here](https://github.com/Kitware/paraview-visualizer/blob/master/pv_visualizer/html/pipeline.py).

**Comment by peterjwilson** [comment]
Thanks a lot Sebastien, your link helped me find the issue: I was appending a GitTree entry instead of re-writing the source. Working example below for reference.

```
from trame.app import get_server
from trame.ui.vuetify import SinglePageLayout
from trame.widgets import vuetify, trame

server = get_server(client_type = "vue2")
state, ctrl = server.state, server.controller

def add_pipeline_node():
    state["pipeline_nodes"] = [{"id": "1", "parent": "0", "visible": 1, "name": "Mesh"},
                               {"id": "2", "parent": "1", "visible": 1, "name": "Contour"}] # this works
    #state["pipeline_nodes"].append({"id": "2", "parent": "1", "visible": 1, "name": "Contour"}) # this doesn't

with SinglePageLayout(server) as layout:
    with layout.content:
        with vuetify.VContainer(fluid=True,classes="pa-0 fill-height",):
            with vuetify.VBtn(icon=True, click=add_pipeline_node):
                vuetify.VIcon("mdi-plus")
            trame.GitTree(sources=("pipeline_nodes",[{"id": "1", "parent": "0", "visible": 1, "name": "Mesh"}]))

if __name__ == "__main__":
    server.start()
```

**Reply by jourdain** [reply, marked as answer]
So the issue you ran into is because you are mutating a state variable in a way that the shared state was not able to detect. A way to handle it without replacing the state variable is to call [dirty](https://trame.readthedocs.io/en/latest/core.state.html#trame_server.state.State.dirty) like stated in the [How to start - forcing state exchange](https://kitware.github.io/trame/guide/intro/getting_started.html#forcing-state-exchange).

In other words, you could also have done

```python
def add_pipeline_node():
    state.pipeline_nodes.append({"id": "2", "parent": "1", "visible": 1, "name": "Contour"})
    state.dirty("pipeline_nodes")
```

**Reply by peterjwilson** [reply]
Yep, that works perfectly. Much appreciated @jourdain!
